opt: internal error: cannot dynamically construct operator unknown

roachtest.sqlsmith/setup=empty/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5753958?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5753958?buildTab=artifacts#/sqlsmith/setup=empty/setting=no-mutations) on master @ [e9ee21860458d997a8155734dc608cfcd050ef24](https://github.com/cockroachdb/cockroach/commits/e9ee21860458d997a8155734dc608cfcd050ef24):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:265,sqlsmith.go:325,test_runner.go:896: error: pq: internal error: cannot dynamically construct operator unknown
		stmt:
		SELECT
			tab_49694.col_85971 AS col_85973, tab_49695.col_85972 AS col_85974
		FROM
			(
				VALUES
					(B'001011010111011101001101010000'),
					(B'110111111101111010001100000101'),
					(B'111101001110111011111000010010')
			)
				AS tab_49694 (col_85971),
			(
				VALUES
					(NULL),
					(
						'0103000080010000000A000000045B1A421F3BE0C10E017E88B054E4C17E4EB457B991F5C1C1A0B294AC48F0C10C79EF72CC3601C2868B29BB89D9F9C1BC8377695EF3F1410D3F6F4EAD7500C2CCFFC2185ED7F84136CAAEC1BE7A0142C3A608961492F5C1A0ECC3187952DC41A82D37003ACEFF415E2E54912F4DF3C190B73359F769F1417A1A3985F63BF5414E806CA3A03CE1C12C9A8D41F82FFC41E24E93FBB52BF341982CA6839637F041636485B0D10402C2C4D77DC0E3A3E941E8B4F5D1A3E5FB41C081683C0651FC4188F2922D8603D2C1B427C5D62565FE41A0167CCCFBF8FD41045B1A421F3BE0C10E017E88B054E4C17E4EB457B991F5C1':::GEOMETRY
					),
					(
						'01040000400400000001010000401E574E96B79AFFC1ACEC12E3F2A2FBC1D4576DC84108E8C101010000400058F5C4FB36A241A0F298F12EF1CF41226D2A4D0BD701420101000040C4BAE5C3C107EF4184747612622C01C2A0E11F74CAD1F9C10101000040C4D473482188F8C19C746E18EAB70142C850FFE78016E8C1':::GEOMETRY
					),
					(
						'01020000800300000028ACD8EC4A63D1414407B759AD1FFF41103E12006D5DC6C1FD13E4E8A84CFDC15E769FC25863F341583C3A7D799AD941685431299525E6C140F8D8D4F268D6417AB6C32EBD7AF6C1':::GEOMETRY
					),
					(st_voronoilines(NULL::GEOMETRY)::GEOMETRY),
					(
						'010500004004000000010200004003000000C087A735F8FBBF416C730FACF74BF141B4D434EE3B9E01C2805C4A7C1F16DFC1A8A42EF70FB5F341C44B2A3F2EEAEC410A58C4B4119EFDC16A3F88B375130042D8C638A7AB690042010200004002000000C4AAF1C29822FF414C1F2DBA89F0E5C120C071CBD166DC41EC9016194D8BE54100AE7EE6784AF741FF7247BD058DFAC101020000400400000034B1A9FA47EBEFC1683C79863B7D0242AC805F2153A8F641C86BC00EE064F7C1D0B43FEA8017FA41BF0D434A4E93F1C1C05F00DC094EDDC1AC6B30EAC56DE9416CE1EAE5E0F7F3C12B1CD2ABB319F7C1322DCDCBA379F14190AE8BE1B4A3D441010200004006000000A0A28AC6367CFE41F0BE3D2C72B4DF412446AB2E4BBBF4C180BDE11B7C77F4410852BEEBA689004220D4984B2274C441409EAC1F1FC4D2417E9B87C12571F4410AB0518A7BB1F841605EF4774480D2C154E57E2961B9F241A4D4BBCD8EB1EAC13CD7F2D9E9D6EBC106708C8E1D8302420289C8B0E4170042F88C487089B1EAC1AE37410C97C3F541E02CFCB01AB0EA41':::GEOMETRY
					)
			)
				AS tab_49695 (col_85972)
		WHERE
			(NOT (tab_49695.col_85972 ~ tab_49695.col_85972))
		LIMIT
			78:::INT8;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17681

--- a/pkg/sql/opt/norm/bool_funcs.go
+++ b/pkg/sql/opt/norm/bool_funcs.go
@@ -37,6 +37,12 @@ func (c *CustomFuncs) NegateComparison(
 	return c.f.DynamicConstruct(negate, left, right).(opt.ScalarExpr)
 }
 
+// CanNegateComparison returns whether the given comparison op can be negated.
+func (c *CustomFuncs) CanNegateComparison(cmp opt.Operator) bool {
+	_, ok := opt.NegateOpMap[cmp]
+	return ok
+}
+
 // FindRedundantConjunct takes the left and right operands of an Or operator as
 // input. It examines each conjunct from the left expression and determines
 // whether it appears as a conjunct in the right expression. If so, it returns

--- a/pkg/sql/opt/norm/rules/bool.opt
+++ b/pkg/sql/opt/norm/rules/bool.opt
@@ -110,16 +110,14 @@ $input
 
 # NegateComparison inverts eligible comparison operators when they are negated
 # by the Not operator. For example, Eq maps to Ne, and Gt maps to Le. All
-# comparisons can be negated except for the JSON comparisons.
+# comparisons can be negated except for the JSON and Geospatial comparisons.
 [NegateComparison, Normalize]
 (Not
     $input:(Comparison $left:* $right:*) &
-        ^(Contains | ContainedBy | JsonExists | JsonSomeExists
-                | JsonAllExists | Overlaps
-        )
+        (CanNegateComparison $op:(OpName $input))
 )
 =>
-(NegateComparison (OpName $input) $left $right)
+(NegateComparison $op $left $right)
 
 # EliminateNot discards a doubled Not operator.
 [EliminateNot, Normalize]

--- a/pkg/sql/opt/norm/testdata/rules/bool
+++ b/pkg/sql/opt/norm/testdata/rules/bool
@@ -424,6 +424,39 @@ select
       ├── NOT (j:5 ?& ARRAY['foo']) [outer=(5), immutable]
       └── NOT (ARRAY[i:2] && ARRAY[1]) [outer=(2), immutable]
 
+# Regression test for #84476 - don't panic when encountering a negated
+# geospatial comparison operator.
+norm expect-not=NegateComparison
+SELECT
+	NULL, t.g
+FROM
+	(VALUES (NULL::GEOMETRY)) AS t (g)
+WHERE
+	(NOT (t.g ~ t.g));
+----
+project
+ ├── columns: "?column?":2 g:1
+ ├── cardinality: [0 - 1]
+ ├── immutable
+ ├── key: ()
+ ├── fd: ()-->(1,2)
+ ├── select
+ │    ├── columns: column1:1
+ │    ├── cardinality: [0 - 1]
+ │    ├── immutable
+ │    ├── key: ()
+ │    ├── fd: ()-->(1)
+ │    ├── values
+ │    │    ├── columns: column1:1
+ │    │    ├── cardinality: [1 - 1]
+ │    │    ├── key: ()
+ │    │    ├── fd: ()-->(1)
+ │    │    └── (NULL,)
+ │    └── filters
+ │         └── NOT (CAST(NULL AS GEOMETRY) ~ CAST(NULL AS GEOMETRY)) [immutable]
+ └── projections
+      └── NULL [as="?column?":2]
+
 # --------------------------------------------------
 # EliminateNot
 # --------------------------------------------------

